plugins/node/opentelemetry-instrumentation-mongodb/src/mongodb.ts

```diff
@@ -0,0 +1,410 @@
+        }
+        this._wrap(moduleExports, 'remove', this._getPatchOperation('remove'));
+        // patch update operation
+        if (isWrapped(moduleExports.update)) {
+          this._unwrap(moduleExports, 'update');
+        }
+        this._wrap(moduleExports, 'update', this._getPatchOperation('update'));
+        // patch other command
+        if (isWrapped(moduleExports.command)) {
+          this._unwrap(moduleExports, 'command');
+        }
+        this._wrap(moduleExports, 'command', this._getPatchCommand());
+        // patch query
+        if (isWrapped(moduleExports.query)) {
+          this._unwrap(moduleExports, 'query');
+        }
+        this._wrap(moduleExports, 'query', this._getPatchFind());
+        // patch get more operation on cursor
+        if (isWrapped(moduleExports.getMore)) {
+          this._unwrap(moduleExports, 'getMore');
+        }
+        this._wrap(moduleExports, 'getMore', this._getPatchCursor());
+        return moduleExports;
+      },
+      unpatch: (moduleExports?: T, moduleVersion?: string) => {
+        if (moduleExports === undefined) return;
+        diag.debug(`Removing internal patch for mongodb@${moduleVersion}`);
+        this._unwrap(moduleExports, 'insert');
+        this._unwrap(moduleExports, 'remove');
+        this._unwrap(moduleExports, 'update');
+        this._unwrap(moduleExports, 'command');
+        this._unwrap(moduleExports, 'query');
+        this._unwrap(moduleExports, 'getMore');
+      },
+    };
+  }
+
+  /** Creates spans for common operations */
+  private _getPatchOperation(operationName: 'insert' | 'update' | 'remove') {
+    const instrumentation = this;
+    return (original: WireProtocolInternal[typeof operationName]) => {
+      return function patchedServerCommand(
+        this: unknown,
+        server: MongoInternalTopology,
+        ns: string,
+        ops: unknown[],
+        options: unknown | Function,
+        callback?: Function
+      ) {
+        const currentSpan = getSpan(context.active());
+        const resultHandler =
+          typeof options === 'function' ? options : callback;
+        if (
+          !currentSpan ||
+          typeof resultHandler !== 'function' ||
+          typeof ops !== 'object'
+        ) {
+          return original.call(
+            this,
+            server,
+            ns,
+            ops,
+            typeof options === 'function' ? callback : options,
+            callback
+          );
+        }
+        const span = instrumentation.tracer.startSpan(
+          `mongodb.${operationName}`,
+          {
+            kind: SpanKind.CLIENT,
+          }
+        );
+        instrumentation._populateAttributes(
+          span,
+          ns,
+          server,
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          operationName !== 'insert' ? (ops[0] as any) : undefined
+        );
+        const patchedCallback = instrumentation._patchEnd(span, resultHandler);
```

Comment:
Not sure here but I think this is problematic.
Callbacks are optional in mongodb APIs (see e.g. http://mongodb.github.io/node-mongodb-native/3.6/api/Collection.html#remove) and if no callback is given a promise is returned. Looks like we may create a callback here which would change behavior for users.

Reply:
From what i saw all the internal here are using callback-based system ([here](https://github.com/mongodb/node-mongodb-native/blob/3.6/lib/core/wireprotocol/write_command.js#L9)) so i believe the promise are handled much higher in the API.